test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj

```diff
@@ -119,6 +119,44 @@
           response (ring-app request)]
       (is (= version-number (get-in response [:headers "X-Puppet-Version"]))))))
 
+(deftest handle-delete-certificate-request!-test
+  (let [settings      (assoc (testutils/ca-sandbox! cadir)
+                             :allow-duplicate-certs true
+                             :autosign false)
+        subject       "foo-agent"
+        csr-stream    (io/input-stream (test-pem-file "foo-agent-csr.pem"))
+        expected-path (ca/path-to-cert-request (:csrdir settings) subject)
+        assert-proper
+          (fn
+            [response code matcher log-level]
+            (do
+             (is (= code (:status response)))
+             (is (re-matches matcher (:body response)))
+             (is (= "text/plain" (get-in response [:headers "Content-Type"])))
+             (is (logged? matcher log-level))))]
+    (logutils/with-test-logging
+      (testing "Successful CSR deletion"
+        (try
+          (handle-put-certificate-request! subject csr-stream settings)
+          (is (true? (fs/exists? expected-path)))
+          (let [response (handle-delete-certificate-request! subject settings)
+                msg-matcher (re-pattern (str "Deleted .* for " subject ".*"))]
+            (is (false? (fs/exists? expected-path)))
+            (assert-proper response 204 msg-matcher :debug))
+          (finally
+            (fs/delete expected-path))))
+      (testing "Attempted deletion of a non-existant CSR"
+        (let [response (handle-delete-certificate-request! subject settings)
+              msg-matcher (re-pattern (str "No cert.* " subject " at .*"))]
+          (is (false? (fs/exists? expected-path)))
+          (assert-proper response 404 msg-matcher :info)))
+      (testing "Error during deletion of a CSR"
+        (let [response
+              (with-redefs [me.raynes.fs/exists? (constantly true)]
```

Comment:
Sorry I spaced responding to your comment about deleting the file. I think the [code branching](https://github.com/puppetlabs/puppet-server/pull/998#discussion-diff-58047866R1000) is such that we only get to the error condition if the file exists but we cannot delete it. In order to trigger it via deletion we'd have to delete the file between lines 1000 and 1001 in pl.ps.certificate-authority and I'm not sure how to do that.

I'm concerned about changing the permissions of a whole directory of static test fixtures for the same reason I'm concerned about `with-redefs`. If I understand it correctly, if the tests are ran in parallel both will cause spurious failures in unrelated tests.

I think if the options are between changing perms on the fixtures directory vs using `with-redefs`, I, naively, have a slight preference to using `with-redefs` since it's in code and I think we have more control over the code than a filesystem the tests run on.

So those are my concerns with the two options mentioned @KevinCorcoran but I'm probably missing something and don't know what I don't know.